File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        vector<int> freq(26, 0);
        for (char c : s) {
            freq[c - 'a']++;
        }
        
        sort(freq.begin(), freq.end());
        
        int maxAllowed = s.size(), deletions = 0;
        
        for (int i = 25; i >= 0 && freq[i] > 0; i--) {
            if (freq[i] > maxAllowed) {
                deletions += freq[i] - maxAllowed;
                freq[i] = maxAllowed;
            }
            maxAllowed = max(0, freq[i] - 1);
        }
        return deletions;
    }
};

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.java
class Solution {
    public int minDeletions(String s) {
        int[] freq = new int[26];
        
        for (char c : s.toCharArray())
            freq[c - 'a']++;
        
        Arrays.sort(freq);
        
        int max = s.length(), deletions = 0;
        
        for (int i = 25; i >= 0; i--) {
            if (freq[i] == 0)
                continue;
            else {
                if (freq[i] > max){
                    deletions += freq[i] - max;
                    freq[i] = max;
                }
                
                max = Math.max(0, freq[i] - 1);
            }
        }
        return deletions;
    }
}